packages/payment-processor/src/payment/erc20-fee-proxy.ts

```diff
@@ -0,0 +1,181 @@
+ */
+export function encodePayErc20FeeRequest(
+  request: ClientTypes.IRequestData,
+  signerOrProvider: Web3Provider | Signer = getProvider(),
+  amount?: BigNumberish,
+  feeAmountOverride?: BigNumberish,
+): string {
+  validateRequest(request, PaymentTypes.PAYMENT_NETWORK_ID.ERC20_FEE_PROXY_CONTRACT);
+  const signer = getSigner(signerOrProvider);
+
+  const tokenAddress = request.currencyInfo.value;
+  const proxyAddress = erc20FeeProxyArtifact.getAddress(request.currencyInfo.network!);
+
+  const { paymentReference, paymentAddress, feeAddress, feeAmount } = getRequestPaymentValues(
+    request,
+  );
+
+  if (!!feeAmount !== !!feeAddress) {
+    throw new Error('Fee amount or not undefined');
+  }
+
+  const amountToPay = getAmountToPay(request, amount);
+  const feeToPay = feeAmountOverride || bigNumberify(feeAmount || 0);
+
+  const proxyContract = Erc20FeeProxyContract.connect(proxyAddress, signer);
+
+  return proxyContract.interface.functions.transferFromWithReferenceAndFee.encode([
+    tokenAddress,
+    paymentAddress,
+    amountToPay,
+    `0x${paymentReference}`,
+    feeToPay,
+    feeAddress || '',
+  ]);
+}
+
+/**
+ * Checks if a given account has the necessary allowance to pay an ERC20 request.
+ * @param request request to pay
+ * @param account account that will be used to pay the request
+ * @param provider the web3 provider. Defaults to Etherscan.
+ */
+export async function hasErc20Approval(
+  request: ClientTypes.IRequestData,
+  account: string,
+  provider: Provider = getNetworkProvider(request),
+): Promise<boolean> {
+  const erc20Contract = ERC20Contract.connect(request.currencyInfo.value, provider);
+  const allowance = await erc20Contract.allowance(
+    account,
+    erc20FeeProxyArtifact.getAddress(request.currencyInfo.network!),
+  );
+  const { feeAmount } = getRequestPaymentValues(request);
+  return allowance.gt(bigNumberify(request.expectedAmount).add(feeAmount || 0));
+}
+
+/**
+ * Processes the approval transaction of the targeted ERC20.
+ * @param request request to pay
+ * @param provider the web3 provider. Defaults to Etherscan.
+ * @param overrides optionally, override default transaction values, like gas.
+ */
+export async function approveErc20(
+  request: ClientTypes.IRequestData,
+  signerOrProvider: Web3Provider | Signer = getProvider(),
+  overrides?: ITransactionOverrides,
+): Promise<ContractTransaction> {
+  const encodedTx = encodeApproveErc20(request, signerOrProvider);
+  const signer = getSigner(signerOrProvider);
+  const tokenAddress = request.currencyInfo.value;
+  const tx = await signer.sendTransaction({
+    data: encodedTx,
+    to: tokenAddress,
+    value: 0,
+    ...overrides,
+  });
+  return tx;
+}
+
+/**
+ * Encodes the approval call, can be used with a Multisig contract.
+ * @param request the request to pay
+ * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ */
+export function encodeApproveErc20(
```

Comment:
maybe this method can be reused from the other PN? 
only the `validateRequest` changes, I believe, and as it's checked on the payment, maybe it's enough?

Reply:
It's also the address. I'll make it more generic on the other PN and reuse.